### ISSUE TITLE
Set Terragrunt defaults in terragrunt.hcl file

### ISSUE THREAD

**Issue by mossad-zika** [gruntwork-io/terragrunt#3572]
This is a duplicate of https://github.com/gruntwork-io/terragrunt/issues/2849

Currently, Terragrunt is able to pass extra arguments and set environment variables for Terraform using hooks or run_cmd() command. However, this does not work for Terragrunt itself.

For example, we constantly have to type terragrunt run-all plan --terragrunt-non-interactive, whereas I would want to set this flag in terragrunt.hcl file (or as an env var TERRAGRUNT_NON_INTERACTIVE=true) and have Terragrunt honour it.

Describe alternatives you've considered

Wrapper for terragrunt executables, such as Makefiles and bash scripts - but those are messy.
Setting TERRAGRUNT_ variables manually on the CLI before running Terragrunt but, again, yet another step and more typing that shouldn't be needed.
Additional context
If there's any way to set Terragrunt defaults within itself, that would be great as the number of options is growing and can make for a messy and long command line.

**Comment by yhakbar** [comment]
Hey @mossad-zika , we've talked about this internally, and we have some designs in mind.

The design that we're considering right now is basically to have a `.terragruntrc.json` file that can go in places you would typically expect it:

1. Current working directory.
2. Root of repository.
3. In `.config` directory at root of repository.
4. Home directory of user.

The file would be parsed, then set Terragrunt flags with names that match the names of flags, so it would be pretty easy to guess how configurations would be set.

In the future there may be other configurations that we'd add to this, but that's our general thought process right now.

Marking the issue as `preserved`, as we plan to address this, but it's just not a top priority right now.

If you would like this issue resolved earlier, consider cutting a pull request to deliver this.

**Comment by BeyondEvil** [comment]
Yes! This!

And please please please use JSON and not yaml 😅

**Comment by yhakbar** [comment]
Something to consider is that this is becoming more important, as we're evaluating changing some defaults in how Terragrunt behaves.

[This flag](https://terragrunt.gruntwork.io/docs/reference/cli-options/#terragrunt-no-destroy-dependencies-check), for example, is not a great default for Terragrunt to have, but is also non-trivial to remove. It's nice for users to have that check happen whenever they try to destroy something, and it's only useful when users _don't know_ that there's a dependency on something they're destroying, so it's not great to opt-in to that functionality. We could have some recommended `.terragruntrc.json` files in the docs for that feature that explicitly enable the check for users that are willing to accept the performance penalty.

**Comment by yhakbar** [comment]
It could look like this:

```json
{
  "flags": [
    {
      "name": "non-interactive",
      "default": true,
    }
  ]
}
```

or this:

```json
{
  "commands": [
    {
      "name": "run",
      "flags" : [
        { "name": "non-interactive", "default": true }
      ]
    }
  ]
}
```

e.g.

```json
{  
  "flags": [
    {
      "name": "non-interactive",
      "default": true,
    }
  ],
  "commands": [
    {
      "name": "run",
      "flags" : [
        { "name": "non-interactive", "default": true }
      ]
    }
  ]
}
```



And we could support YAML.

**Comment by captainjack64** [comment]
Any updates on this feature? It would be incredibly helpful.

**Comment by yhakbar** [comment]
@captainjack64 

We haven't had time to prioritize this, and it doesn't seem like the community is that interested in the feature. If you'd like to, you can try to convince more of the community to share their feedback that they'd want this feature and any thoughts on the kind of design they'd like for it. If we see that the community is really clamoring for this, we can look to set aside some time to get it built out.

**Comment by zhemaituk** [comment]
@yhakbar my use case:

Currently I have to rely on readme and each engineer has to remember to read it and manually configure/update their local env. But we all know engineers don't read documentation, and especially don't read it again if it changes :).

It would be amazing to have a way (json, yaml, htl, any file format) to specify the default config for the project. If someone needs to override it - there are current ways (env vars or cli flags). But if someone just runs the command without any overrides - it behaves the right intended/default way.

An example from the readme:

Add to `~/.zshrc`:

```shell
# TODO: Move to project config once https://github.com/gruntwork-io/terragrunt/issues/3572 is addressed
# Using shared provider cache significantly decreases disc space usage and speeds up initialization. 150+ GB down to ~1 GB.
export TG_PROVIDER_CACHE="1"
export TG_PROVIDER_CACHE_DIR="..."
```

**Comment by pllhx** [comment]
Any news on this? I'd really like to have this! A single central config file to set TG flags would be so much better than a bunch of environment variables I have set in either a Makefile, or .bashrc/.profile etc. Especially if I can set it at the project level and commit it to git for my entire team to use!

**Comment by tomektom** [comment]
@zhemaituk you definitely can simplify this by using [direnv](https://direnv.net) or other similar solution and tell them how to configure it properly (and you need do it only once). And keep in repo your `.envrc` with shared config in repo 😉
